### DISCUSSION TITLE
Project Status?

### DISCUSSION THREAD

**Discussion by prologic** [aklinker1/miasma#19] [Q&A]
Hey! 👋

Are you still working on this project?

Where do you currently use this?

Do you have a roadmap or future plans?

Kind regards

James

**Comment by aklinker1** [comment]
I work on it every now and then between my other projects. Recently I started converting it from rest to graphql, and improving the architecture/code quality. I wrote this when I was ignorant of the right way to write go code.

I do use it personally, but I moved recently and my home network is in shambles lol.

Here's a rough roadmap:

1. Finish graphql refactor (see `refactor` branch)
2. Prevent CLI from timing out when downloading large docker images
3. Add a UI as an alternative to the CLI
4. Prettify the CLI output (add colors, loading indicators, etc)

I'd say end of year would be a reasonable estimate for all that work.

**Reply by aklinker1** [reply]
I'll create issues for all those things later today and put them in a milestone for people to track.

**Reply by aklinker1** [reply]
https://github.com/aklinker1/miasma/milestone/4

**Reply by prologic** [reply]
That's great to hear 👌 Couple of follow-up questions and comments:

1. Why GraphQL? What problems are you hoping to solve by moving from REST to GraphQL?
2. The backend currently has critical bugs that are probably of a higher priority/criticality to fix IHMO, was doing some testing last night and wasn't able to get very far. I don't think you publish a Docker image either at the moment -- As I had to build it myself.
3. Speaking of building the code, I was unable to build the codebase as-is with `go build` as you depend on SQLite, and I don't normally carry this C dependency around with me on various systems I develop/test on. Any chance you can switch this dependency out with something like BoltDB, Badger or even my own [Bitcask](https://pkg.go.dev/git.mills.io/prologic/bitcask)?

That's all 🙇‍♂️

**Reply by aklinker1** [reply]
I should have said this first, but thanks for showing interest in my project! Now onto your questions...

> 1. **Why GraphQL? What problems are you hoping to solve by moving from REST to GraphQL?**

In the project, there are two data sources: docker and the local database. Since docker can be quite slow to respond with data, GraphQL can help speed up responses if queries only request fields from one of those data sources, like the local database. Relatively speaking, it's still fast enough, so a REST API would probably be fine here, which is why I chose REST in the first place.

The other reason is that the GraphQL library I use, [gqlgen](https://gqlgen.com/), generates a lot of code that I would otherwise have to write manually (models, enums, docs, and field validation being the big ones). Big time save on that front.

TBH though, I chose it because I'm much more motivated to work on GraphQL APIs compared to REST APIs. I just enjoy them more. One of the goals of the refactor is to make the code more modular so I can swap things out as needed. I could very easily swap out GraphQL for REST, or support both if needed in the future.

> 2. **The backend currently has critical bugs that are probably of a higher priority/criticality to fix IHMO, was doing some testing last night and wasn't able to get very far. I don't think you publish a Docker image either at the moment -- As I had to build it myself.**

Most these bugs stem from poorly structured code, which will be solved in the refactor.

I do publish a docker image, but docker hub deletes images if new ones aren't published regularly, which is why you couldn't find one :shrug: 

> 3. **Speaking of building the code, I was unable to build the codebase as-is with `go build` as you depend on SQLite, and I don't normally carry this C dependency around with me on various systems I develop/test on. Any chance you can switch this dependency out with something like BoltDB, Badger or even my own [Bitcask](https://pkg.go.dev/git.mills.io/prologic/bitcask)?**

This is my bad for not documenting the build process and tooling requirements, but [you should be running `pnpm build:server` instead](https://github.com/aklinker1/miasma/blob/c193b9a6c65bb54af9f06abe55d571793358c146/package.json#L5). It builds the code in a docker container that contains the C dependency. Checkout the scripts in the `package.json` for more information. All you need installed is `docker`, `node`, and `pnpm` (`npm i -g pnpm`). Optionally, `go` if your editor requires it, like VS Code.

> Normally I use a `Makefile` to store my build scripts, but since I added a frontend, I needed a package.json and didn't want both of them. I haven't enjoyed just the `package.json`, so I might switch back to the `Makefile`

It still may not build though. I'm not really concerned about this while I'm in the middle of the refactor and this project isn't used enough by other people that I feel I need to maintain it to the same standards as other OSS like docker or K8s.

As for the database, SQLite it's a battle tested database and I can easily find answers to my questions. I have lots of experience working with SQL database, so it was a comfortable solution. I also love the `database/sql` standard library, so being able to use that was a win for me as well.

Right now, I mostly access data via it's ID, so key/value DBs like you mentioned could work, but I don't want to limit myself to a database that doesn't support larger queries if I need them in the future.

However, I could be talked into using a different database, specifically because SQLite adds a minute or more to the build process (I prefer <1s build times). I've gotten around this by having a [specific step for this in my dockerfile that can be cached](https://github.com/aklinker1/miasma/blob/c193b9a6c65bb54af9f06abe55d571793358c146/Dockerfile#L38), but when that cache isn't hit, it's quite annoying and really slows me down.
